fix: Update ObjectStoragePathFactory to allow for custom validation

frifriSF59 opened the pull request:
## What

Update the `ObjectStoragePathFactory` to allow us to pass custom transform logic on the namespace and stream name.
This new logic is optional and should not impact any of the current connectors.

The trigger for this change is: https://airbytehq.sentry.io/issues/6690125303/?query=is%3Aunresolved%20Source%20URI%20must%20not%20contain%20the%20%27%2C%27&stream_index=0
 
## Can this PR be safely reverted and rolled back?
- [X] YES 💚
- [ ] NO ❌

frifriSF59 committed:
Update ObjectStoragePathFactory to allow for custom validation

# Conflicts:
#	airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ObjectStoragePathFactory.kt

frifriSF59 committed:
dont need that

edgao commented on airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/load/write/BasicFunctionalityIntegrationTest.kt:
```diff
@@ -744,6 +744,7 @@ abstract class BasicFunctionalityIntegrationTest(
                     makeStream("stream_name_with_operator+1"),
                     makeStream("stream_name_with_numbers_123"),
                     makeStream("1stream_with_a_leading_number"),
+                    makeStream("c'est une belle histoire!./?,<}[=+-_$%^&!@#*&(~`"),
```
```suggestion
                    makeStream("c'est une belle histoire,./<>?'\";[]\\:{}|`~!@#\$%^&*()_+-="),
```
(just adding more special chars 😅 )

edgao commented on airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/command/object_storage/ObjectStoragePathConfiguration.kt:
```diff
@@ -8,6 +8,7 @@ data class ObjectStoragePathConfiguration(
     val prefix: String,
     val pathPattern: String?,
     val fileNamePattern: String?,
+    val resolveNamesMethod: ((String) -> String)? = null
```
I think this works? then we can avoid the nullable transform (which is kind of misleading, since you might expect `null` to be a noop transform rather than the actual `s3chars` thing) + have the ObjectStoragePathFactory call the method directly instead of wrapping in `transformName`
```suggestion
    val resolveNamesMethod: ((String) -> String) = { Transformations.toS3SafeCharacters(it) }
```

frifriSF59 committed:
RC3

# Conflicts:
#	airbyte-integrations/connectors/destination-bigquery/metadata.yaml
#	docs/integrations/destinations/bigquery.md

frifriSF59 committed:
Update that

frifriSF59 committed:
Format

frifriSF59 committed:
Private

edgao commented on airbyte-integrations/connectors/destination-bigquery/src/main/kotlin/io/airbyte/integrations/destination/bigquery/write/bulk_loader/BigQueryBulkLoaderConfiguration.kt:
```diff
@@ -75,6 +76,7 @@ data class BigqueryBulkLoadConfiguration(
                 pathPattern =
                     "\${NAMESPACE}/\${STREAM_NAME}/\${YEAR}/\${MONTH}/\${DAY}/\${HOUR}/\${UUID}",
                 fileNamePattern = "{date}_{timestamp}_{part_number}{format_extension}",
+                resolveNamesMethod = { name -> Transformations.toAlphanumericAndUnderscore(name) }
```
slightly cleaner
```suggestion
                resolveNamesMethod = { Transformations.toAlphanumericAndUnderscore(it) }
```

edgao commented on airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ObjectStoragePathFactory.kt:
```diff
@@ -103,6 +104,11 @@ class ObjectStoragePathFactory(
             fileFormatExtension ?: compressionExtension
         }
 
+    private val PATH_VARIABLES = getPathVariables(pathConfig)
```
nit: rename these to `pathVariables` / `pathVariablesStreamConstant`

edgao commented on airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ObjectStoragePathFactory.kt:
```diff
@@ -161,6 +167,55 @@ class ObjectStoragePathFactory(
+    private fun getPathVariables(pathConfig: ObjectStoragePathConfiguration): List<PathVariable> {
+        fun transformName(name: String): String {
+            return when {
+                pathConfig.resolveNamesMethod != null -> pathConfig.resolveNamesMethod.invoke(name)
```
```suggestion
                pathConfig.resolveNamesMethod != null -> pathConfig.resolveNamesMethod(name)
```

frifriSF59 committed:
oops. that got removed

frifriSF59 committed:
Move that

frifriSF59 committed:
Variable rename
